src/dahlia/utils.py

```diff
@@ -64,3 +67,224 @@ def _with_marker(marker: str) -> list[Pattern]:
+    def to_4(self):
+        ...
+
+    @abstractmethod
+    def to_8(self):
+        ...
+
+
+class _ANSI_3(_ANSI):
+    def __init__(self, _ansi: list[str], old_ansi: str) -> None:
+        self.old_ansi = old_ansi
+
+    def to_3(self):
+        return self.old_ansi
+
+    def to_4(self):
+        return self.old_ansi
+
+    def to_8(self):
+        return self.old_ansi
+
+
+class _ANSI_4(_ANSI):
+    def __init__(self, ansi: list[str], old_ansi: str) -> None:
+        if ansi[0] == 1:
+            self.bold = True
+            color = int(ansi[1])
+        else:
+            color = int(ansi[0])
+
+        self.color = color % 10
+
+        if 40 <= color <= 47 or 100 <= color <= 107:
+            self.background = True
+
+        self.old_ansi = old_ansi
+
+    def to_3(self):
+        color = self.color + 30
+
+        if self.background:
+            color += 10
+
+        if self.bold:
+            return f"\x1b[1;{color}m"
+
+        return f"\x1b[{color}m"
+
+    def to_4(self):
+        return self.old_ansi
+
+    def to_8(self):
+        return self.old_ansi
+
+
+class _ANSI_8(_ANSI):
+    def __init__(self, ansi: list[str], old_ansi: str) -> None:
+        self.color = int(ansi[2])
+        self.background = ansi[0] == "48"
+        self.old_ansi = old_ansi
+
+    def to_3(self) -> str:
+        eight = _quantize_8_bit(self.color, to=3)
+
+        if isinstance(eight, int):
+            return f"\x1b[{eight + (10 if self.background else 0)}m"
+        else:
+            return _estimate_ansi_color(eight, _COLORS_3, background=self.background)
+
+    def to_4(self) -> str:
+        eight = _quantize_8_bit(self.color, to=4)
+
+        if isinstance(eight, int):
+            return f"\x1b[{eight + (10 if self.background else 0)}m"
+        else:
+            return _estimate_ansi_color(eight, _COLORS_4, background=self.background)
+
+    def to_8(self) -> str:
+        return self.old_ansi
+
+
+class _ANSI_24(_ANSI):
+    def __init__(self, ansi: list[str], _old_ansi: str) -> None:
+        self.rgb = (int(ansi[2]), int(ansi[3]), int(ansi[4]))
+        self.background = ansi[0] == "48"
+
+    def to_3(self) -> str:
+        return _estimate_ansi_color(self.rgb, _COLORS_3, background=self.background)
+
+    def to_4(self) -> str:
+        return _estimate_ansi_color(self.rgb, _COLORS_4, background=self.background)
+
+    def to_8(self) -> str:
+        r, g, b = self.rgb
+
+        color = 36 * (r // 51) + 6 * (g // 51) + (b // 51)
+
+        return f"\x1b[{48 if self.background else 38};5;{color + 16}m"
+
+
+def _build_ansi(old_ansi: str) -> _ANSI:
+    ansi = old_ansi.split(";")
+
+    ansi[0] = ansi[0].removeprefix("\x1b[")
+    ansi[-1] = ansi[-1].removesuffix("m")
+
+    if len(ansi) < 3:
+        if ansi[0] == 1:
+            color = int(ansi[1])
+        else:
+            color = int(ansi[0])
+
+        if color < 90:
+            return _ANSI_3(ansi, old_ansi)
+        else:
+            return _ANSI_4(ansi, old_ansi)
```

Comment:
Thoughts on sth like this?
```suggestion
       (_ANSI_3 if color < 90 else _ANSI_4)(ansi, old_ansi)
```

Reply:
ugly ngl